pandas/core/groupby/groupby.py

```diff
@@ -1531,9 +1542,21 @@ def size(self) -> FrameOrSeriesUnion:
 
     @doc(_groupby_agg_method_template, fname="sum", no=True, mc=0)
     def sum(self, numeric_only: bool = True, min_count: int = 0):
-        return self._agg_general(
-            numeric_only=numeric_only, min_count=min_count, alias="add", npfunc=np.sum
-        )
+
+        # If we are grouping on categoricals we want unobserved categories to
+        # return zero, rather than the default of NaN which the reindexing in
+        # _agg_general() returns. We set self.observed=True for the call to
+        # _agg_general() to avoid it reindexing. We then reindex below with
+        # fill_value=0. See GH 31422
+        with _observed_is_true(self):
+            result = self._agg_general(
+                numeric_only=numeric_only,
+                min_count=min_count,
+                alias="add",
+                npfunc=np.sum,
+            )
+
+        return self._reindex_output(result, fill_value=0)
```

Comment:
i am concerned this is going to fail on duplicated output

i guess if we don't have test dailies then this is ok

but can we not just .fillna(0)?

Reply:
At this point, `result` will never contain the missing categories (hence no risk of duplicated output). 

The context manager above sets `self.observed=True`, so the call to `_reindex_output()` that occurs within `._agg_general()` does nothing; it just immediately returns the result with no missing categories added. See here:

https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/groupby.py#L2691

We can't use `.fillna(0)` because we still need to add the missing categories, which is done by `._reindex_output()`.